plugin/pulpcore/plugin/stages/declarative_version.py

```diff
@@ -103,11 +110,11 @@ def pipeline_stages(self, new_version):
             list: List of :class:`~pulpcore.plugin.stages.Stage` instances
 
         """
-        return [
-            self.first_stage,
-            QueryExistingArtifacts(), ArtifactDownloader(), ArtifactSaver(),
-            QueryExistingContentUnits(), ContentUnitSaver(),
-        ]
+        pipeline = [self.first_stage]
+        if self.download_artifacts:
+            pipeline.extend([QueryExistingArtifacts(), ArtifactDownloader(), ArtifactSaver()])
```

Comment:
Wondering whether we should still have `QueryExistingArtifacts()` in the non-download case. I think this makes sense for on demand, we might not want to do it for cache only(?).

Reply:
From a correctness interpretation of the name `download_artifacts` not including `QueryExistingArtifacts` is probably wrong. Query+Replace of Artifacts in the pipeline is not the same as downloading. So I see what you are saying.

My concern with adding it is the extra runtime cost; from our testing AIUI that stage adds a non-trivial amount of extra time to the overall runtime from the DB querying when processing say 70K Artifacts.

I think we should either do 1 of a few things.

a) Add back QueryExistingArtifacts and accept the runtime outcome
b) s/download_artifacts/handle_artifacts/ so that the name is more correct.
c) another idea

Reply:
Is there a way to find out later that downloading a remote artifact is not necessary (e.g. when the artifact is requested; I don't know the full picture) Then, not replacing artifacts here would not hurt too much.

Reply:
I'm leaning toward leaving it the way it is.  When artifacts are not downloaded they should not be query/replaced in the pending-artifact either.  Mainly because the replacement isn't deterministic (may or may not be replaced) so logic further down the pipeline should not rely on it.  Leaving `QueryExistingArtifacts` in the pipeline without the downloading stage seems to add overhead  without adding any value.

Reply:
As for the renaming (b), I'm not convinced it add clarity.  I'm not a fan of names that include _"handle"_ because it's such a vague term.

Reply:
I agree the handle name is bad. What about `include_artifact_stages=True` and then it would include/disinclude all 3 of them?

Reply:
What if the core, or a plugin includes an artifact related stage (in the future) that needs to be left in the pipeline even when artifacts are not downloaded?  This naming assumes that will never happen. 

The name `download_artifacts` is toggling a specific behavior that has the side-effect of omitting some of the stages.  The renaming to `include_artifact_stages` changes the toggled behavior to something that has the side-effect of toggling downloading which seems less clear.  I think.

Reply:
Sounds good. I'm leaving the 3 stages being omitted and the name as is.